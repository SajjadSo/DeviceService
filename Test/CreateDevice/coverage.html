
<!DOCTYPE html>
<html>
	<head>
		<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
		<style>
			body {
				background: black;
				color: rgb(80, 80, 80);
			}
			body, pre, #legend span {
				font-family: Menlo, monospace;
				font-weight: bold;
			}
			#topbar {
				background: black;
				position: fixed;
				top: 0; left: 0; right: 0;
				height: 42px;
				border-bottom: 1px solid rgb(80, 80, 80);
			}
			#content {
				margin-top: 50px;
			}
			#nav, #legend {
				float: left;
				margin-left: 10px;
			}
			#legend {
				margin-top: 12px;
			}
			#nav {
				margin-top: 10px;
			}
			#legend span {
				margin: 0 5px;
			}
			.cov0 { color: rgb(192, 0, 0) }
.cov1 { color: rgb(128, 128, 128) }
.cov2 { color: rgb(116, 140, 131) }
.cov3 { color: rgb(104, 152, 134) }
.cov4 { color: rgb(92, 164, 137) }
.cov5 { color: rgb(80, 176, 140) }
.cov6 { color: rgb(68, 188, 143) }
.cov7 { color: rgb(56, 200, 146) }
.cov8 { color: rgb(44, 212, 149) }
.cov9 { color: rgb(32, 224, 152) }
.cov10 { color: rgb(20, 236, 155) }

		</style>
	</head>
	<body>
		<div id="topbar">
			<div id="nav">
				<select id="files">
				
				<option value="file0">DeviceService\Test\CreateDevice\main.go (88.9%)</option>
				
				</select>
			</div>
			<div id="legend">
				<span>not tracked</span>
			
				<span class="cov0">not covered</span>
				<span class="cov8">covered</span>
			
			</div>
		</div>
		<div id="content">
		
		<pre class="file" id="file0" style="display: none">package main

import (
        "DeviceService/Responses"
        "DeviceService/Types"
        "encoding/json"
        "errors"
        "os"

        "github.com/aws/aws-lambda-go/events"
        "github.com/aws/aws-lambda-go/lambda"
        "github.com/aws/aws-sdk-go/aws"
        "github.com/aws/aws-sdk-go/aws/session"
        "github.com/aws/aws-sdk-go/service/dynamodb"
        "github.com/aws/aws-sdk-go/service/dynamodb/dynamodbattribute"
)

// sess Global variable, holds initialized session value
var sess *session.Session

// DevicesTable Global variable, holds devices table name value
var DevicesTable *string

// CheckMissingFields function: check if any fields of request payload is missing
func CheckMissingFields(PayloadJSON map[string]interface{}) error <span class="cov8" title="1">{
        Flag := false
        ErrorText := "Missing fields:"

        if PayloadJSON["id"] == nil </span><span class="cov8" title="1">{
                Flag = true
                ErrorText += " 'id' "
        }</span>
        <span class="cov8" title="1">if PayloadJSON["deviceModel"] == nil </span><span class="cov8" title="1">{
                Flag = true
                ErrorText += " 'deviceModel' "
        }</span>
        <span class="cov8" title="1">if PayloadJSON["name"] == nil </span><span class="cov8" title="1">{
                Flag = true
                ErrorText += " 'name' "
        }</span>
        <span class="cov8" title="1">if PayloadJSON["note"] == nil </span><span class="cov8" title="1">{
                Flag = true
                ErrorText += " 'note' "
        }</span>
        <span class="cov8" title="1">if PayloadJSON["serial"] == nil </span><span class="cov8" title="1">{
                Flag = true
                ErrorText += " 'serial' "
        }</span>

        // If any fields are missing, return error
        <span class="cov8" title="1">if Flag == true </span><span class="cov8" title="1">{
                return errors.New(ErrorText)
        }</span>

        <span class="cov8" title="1">return nil</span>
}

// BindPayloadToDevice function: Bind request json payload to device object
func BindPayloadToDevice(PayloadJSON map[string]interface{}) types.Device <span class="cov8" title="1">{
        return types.Device{
                ID:          PayloadJSON["id"].(string),
                DeviceModel: PayloadJSON["deviceModel"].(string),
                Name:        PayloadJSON["name"].(string),
                Note:        PayloadJSON["note"].(string),
                Serial:      PayloadJSON["serial"].(string),
        }
}</span>

// dbInsertDevice function: Insert request payload to database
func dbInsertDevice(device types.Device) error <span class="cov8" title="1">{
        // Create DynamoDB client
        db := mockDynamoDBClient{dynamodb.New(sess)}

        // Convert Device object to dynamodb AttributeValues
        dbDevice, err := dynamodbattribute.MarshalMap(device)
        if err != nil </span><span class="cov0" title="0">{
                return err
        }</span>

        // Insert dbDevice into DevicesTable
        <span class="cov8" title="1">_, err = db.PutItem(&amp;dynamodb.PutItemInput{
                Item:      dbDevice,
                TableName: DevicesTable,
        })

        if err != nil </span><span class="cov0" title="0">{
                return err
        }</span>

        <span class="cov8" title="1">return nil</span>
}

// Handler function: persists the posted json Device and returns it
// back to the client as an acknowledgement
func Handler(req events.APIGatewayProxyRequest) (events.APIGatewayProxyResponse, error) <span class="cov8" title="1">{
        // Unmarshal request payload, and check if payload is valid
        var PayloadJSON map[string]interface{}
        if err := json.Unmarshal([]byte(req.Body), &amp;PayloadJSON); err != nil </span><span class="cov8" title="1">{
                return responses.ErrBadRequest(errors.New("")), nil
        }</span>

        // Check request payload for missing fields
        <span class="cov8" title="1">if err := CheckMissingFields(PayloadJSON); err != nil </span><span class="cov8" title="1">{
                return responses.ErrBadRequest(err), nil
        }</span>

        // Bind request payload to device struct
        <span class="cov8" title="1">device := BindPayloadToDevice(PayloadJSON)

        // Insert device object to DB
        if err := dbInsertDevice(device); err != nil </span><span class="cov0" title="0">{
                return responses.ErrInternalServer(), nil
        }</span>

        // Return inserted device to response
        <span class="cov8" title="1">deviceJSON, err := json.Marshal(device)
        if err != nil </span><span class="cov0" title="0">{
                return responses.ErrInternalServer(), nil
        }</span>

        <span class="cov8" title="1">return responses.SuccessCreated(deviceJSON), nil</span>
}

func main() <span class="cov0" title="0">{
        lambda.Start(Handler)
}</span>

func init() <span class="cov8" title="1">{
        // Initialize a session in AWS_REGION that the SDK will use to load
        region := os.Getenv("AWS_REGION")
        sess = session.Must(session.NewSession(&amp;aws.Config{Region: &amp;region}))

        // Get Devices table name
        DevicesTable = aws.String(os.Getenv("DEVICES_TABLE"))
}</span>
</pre>
		
		</div>
	</body>
	<script>
	(function() {
		var files = document.getElementById('files');
		var visible;
		files.addEventListener('change', onChange, false);
		function select(part) {
			if (visible)
				visible.style.display = 'none';
			visible = document.getElementById(part);
			if (!visible)
				return;
			files.value = part;
			visible.style.display = 'block';
			location.hash = part;
		}
		function onChange() {
			select(files.value);
			window.scrollTo(0, 0);
		}
		if (location.hash != "") {
			select(location.hash.substr(1));
		}
		if (!visible) {
			select("file0");
		}
	})();
	</script>
</html>
