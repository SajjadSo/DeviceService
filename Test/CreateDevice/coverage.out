mode: set
DeviceService\Test\CreateDevice\main.go:19.67,23.30 3 1
DeviceService\Test\CreateDevice\main.go:27.2,27.39 1 1
DeviceService\Test\CreateDevice\main.go:31.2,31.32 1 1
DeviceService\Test\CreateDevice\main.go:35.2,35.32 1 1
DeviceService\Test\CreateDevice\main.go:39.2,39.34 1 1
DeviceService\Test\CreateDevice\main.go:45.2,45.18 1 1
DeviceService\Test\CreateDevice\main.go:49.2,49.12 1 1
DeviceService\Test\CreateDevice\main.go:23.30,26.3 2 1
DeviceService\Test\CreateDevice\main.go:27.39,30.3 2 1
DeviceService\Test\CreateDevice\main.go:31.32,34.3 2 1
DeviceService\Test\CreateDevice\main.go:35.32,38.3 2 1
DeviceService\Test\CreateDevice\main.go:39.34,42.3 2 1
DeviceService\Test\CreateDevice\main.go:45.18,47.3 1 1
DeviceService\Test\CreateDevice\main.go:53.75,61.2 1 1
DeviceService\Test\CreateDevice\main.go:64.48,74.16 5 1
DeviceService\Test\CreateDevice\main.go:79.2,85.16 3 1
DeviceService\Test\CreateDevice\main.go:89.2,89.12 1 1
DeviceService\Test\CreateDevice\main.go:74.16,76.3 1 0
DeviceService\Test\CreateDevice\main.go:85.16,87.3 1 0
DeviceService\Test\CreateDevice\main.go:94.89,97.71 2 1
DeviceService\Test\CreateDevice\main.go:102.2,102.56 1 1
DeviceService\Test\CreateDevice\main.go:107.2,110.47 2 1
DeviceService\Test\CreateDevice\main.go:115.2,116.16 2 1
DeviceService\Test\CreateDevice\main.go:120.2,120.50 1 1
DeviceService\Test\CreateDevice\main.go:97.71,99.3 1 1
DeviceService\Test\CreateDevice\main.go:102.56,104.3 1 1
DeviceService\Test\CreateDevice\main.go:110.47,112.3 1 0
DeviceService\Test\CreateDevice\main.go:116.16,118.3 1 0
DeviceService\Test\CreateDevice\main.go:123.13,125.2 1 0
